### PR TITLE
[FINERACT-1678] business step configuration fix

### PR DESCRIPTION
**Pull request by taskain7**
## Description

Fix for the business step order configuration.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/cob/service/ConfigJobParameterServiceImpl.java [hunk 1]
```diff
@@ -59,15 +60,22 @@ public JobBusinessStepConfigData getBusinessStepConfigByJobName(String jobName)
...
-        List<BatchBusinessStep> batchBusinessSteps = batchBusinessStepRepository.findAllByJobName(jobName);
-        businessSteps.forEach(newBusinessStepConfig -> {
-            BatchBusinessStep batchBusinessStep = batchBusinessSteps.stream()
-                    .filter(oldBusinessStepConfig -> oldBusinessStepConfig.getStepName().equals(newBusinessStepConfig.getStepName()))
-                    .findFirst().orElseThrow(BusinessStepNotBelongsToJobException::new);
-            batchBusinessStep.setStepName(newBusinessStepConfig.getStepName());
-            batchBusinessStep.setStepOrder(newBusinessStepConfig.getOrder());
-            batchBusinessStepRepository.save(batchBusinessStep);
-        });
+        List<String> availableBusinessStepNames = getAvailableBusinessStepsByJobName(BusinessStepCategory.LOAN.name())
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/service/ConfigJobParameterServiceImpl.java:63]: Please merge together the "availableBusinessStepNames" and "notValidBusinessStepNames" filter

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/BusinessConfigurationApiTest.java [hunk 2]
```diff
@@ -96,10 +96,8 @@ public void shouldUpdateStepOrder() {
                 .filter(businessStep -> APPLY_CHARGE_TO_OVERDUE_LOANS.equals(businessStep.getStepName())) //
                 .findFirst().get().getOrder();
         assertEquals(originalOrder + 1, newOrder);
-
-        List<BusinessStep> originalBusinessSteps = getBusinessSteps(originalOrder);
         BusinessStepConfigurationHelper.updateBusinessStepOrder(requestSpec, responseSpec, LOAN_JOB_NAME,
-                BusinessStepConfigurationHelper.toJsonString(originalBusinessSteps));
+                BusinessStepConfigurationHelper.toJsonString(originalStepConfig.getBusinessSteps()));
     }
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/BusinessConfigurationApiTest.java:127]: Please add the following test cases:
1.
- Remove all the steps from the "Loan_job_name"
- Check there is no business steps configured for the loan
- Add "Apply charge to overdue_loans" step
- Check there is 1 business step configured with order 1
- Add "Delinquency Tags business step" to be the first step
- Check there is 2 business steps and the order is correct
- Remove Delinquency Tags business step
- Check there is 1 business step configured with order 1
2. 
- Add a business steps which does not supported by the application -> Correct error flow occurred

File: fineract-provider/src/main/java/org/apache/fineract/cob/domain/BatchBusinessStep.java [hunk 3]
```diff
@@ -33,6 +33,7 @@
 public class BatchBusinessStep extends AbstractPersistableCustom {
 
     @Column(name = "job_name", nullable = false)
+    @Setter
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/domain/BatchBusinessStep.java]: Minor: You can put the "@Setter" on the class.

File: fineract-provider/src/main/java/org/apache/fineract/cob/service/ConfigJobParameterServiceImpl.java (hunk 1 above)
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/service/ConfigJobParameterServiceImpl.java]: Performance improvement: You make this class to implement InitializingBean. In the afterPropertiesSet method can call the "getAvailableBusinessStepsByJobName" method and save the response at class level. The (application wise) available business step implementation will not be changed during runtime.

File: fineract-provider/src/main/java/org/apache/fineract/cob/service/ConfigJobParameterServiceImpl.java [hunk 4]
```diff
@@ -59,15 +60,22 @@ public JobBusinessStepConfigData getBusinessStepConfigByJobName(String jobName)
...
+                .getAvailableBusinessSteps().stream().map(BusinessStepDetail::getStepName).toList();
+        List<String> notValidBusinessStepNames = businessSteps.stream().map(BusinessStep::getStepName)
+                .filter(businessStepName -> !availableBusinessStepNames.contains(businessStepName)).toList();
+        if (notValidBusinessStepNames.isEmpty()) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/service/ConfigJobParameterServiceImpl.java:77]: Is it allowed to remove all of the business steps for a job?

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/BusinessConfigurationApiTest.java [hunk 5]
```diff
@@ -82,32 +85,57 @@ public void shouldReturnApplyChargeToOverdueLoanStepConfigByJobCategory() {
...
         BusinessStepConfigurationHelper.updateBusinessStepOrder(requestSpec, responseSpec, LOAN_JOB_NAME,
                 BusinessStepConfigurationHelper.toJsonString(requestBody));
 
         JobBusinessStepConfigData newStepConfig = BusinessStepConfigurationHelper.getConfiguredBusinessStepsByJobName(requestSpec,
                 responseSpec, LOAN_JOB_NAME);
-        Long newOrder = newStepConfig.getBusinessSteps().stream() //
-                .filter(businessStep -> APPLY_CHARGE_TO_OVERDUE_LOANS.equals(businessStep.getStepName())) //
-                .findFirst().get().getOrder();
-        assertEquals(originalOrder + 1, newOrder);
+        assertEquals(1, newStepConfig.getBusinessSteps().size());
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/BusinessConfigurationApiTest.java:98]: @taskain7 Please test whether the order of the steps are correct

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/BusinessConfigurationApiTest.java [hunk 6]
```diff
@@ -82,32 +85,57 @@ public void shouldReturnApplyChargeToOverdueLoanStepConfigByJobCategory() {
...
+                BusinessStepConfigurationHelper.toJsonString(requestBody));
+        newStepConfig = BusinessStepConfigurationHelper.getConfiguredBusinessStepsByJobName(requestSpec, responseSpec, LOAN_JOB_NAME);
+        assertEquals(2, newStepConfig.getBusinessSteps().size());
+
+        requestBody.remove(1);
+        BusinessStepConfigurationHelper.updateBusinessStepOrder(requestSpec, responseSpec, LOAN_JOB_NAME,
+                BusinessStepConfigurationHelper.toJsonString(requestBody));
+
+        newStepConfig = BusinessStepConfigurationHelper.getConfiguredBusinessStepsByJobName(requestSpec, responseSpec, LOAN_JOB_NAME);
+        assertEquals(1, newStepConfig.getBusinessSteps().size());
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/BusinessConfigurationApiTest.java:121]: Please test whether the order of the step are correct (=1)

**Review by adamsaghy — CHANGES_REQUESTED** [review]
@taskain7 Good work! Just a minor issue... please take a look and after it got fixed it can go in.

**Review by adamsaghy — COMMENTED** [review]
LGTM

**Review by vidakovic — APPROVED** [review]
LGTM

**Pull request merged**